File: lib/youtube_fetcher.dart
import 'package:youtube_explode_dart/youtube_explode_dart.dart';
import 'dart:convert';
import 'dart:io';

/// A YouTube fetcher with an in-memory LRU stream URL cache
/// to dramatically speed up repeated playback requests.
class YouTubeFetcher {
  static final YoutubeExplode _yt = YoutubeExplode();

  // LRU cache: insertion-ordered map, evict first entry when full
  static final _streamCache = <String, String>{};
  static const _cacheMaxSize = 25;

  /// Search for tracks and return a list of results
  Future<List<Video>> searchTracks(String query) async {
    try {
      final searchList = await _yt.search.search(query);
      if (searchList.isNotEmpty) {
        return searchList.take(20).toList();
      }
    } catch (e) {
      debugLog('Library search failed, attempting custom fallback scraper: $e');
    }
    return _customSearchScrape(query);
  }

  Future<List<Video>> _customSearchScrape(String query) async {
    final client = HttpClient();
    try {
      final uri = Uri.parse('https://www.youtube.com/youtubei/v1/search');
      final request = await client.postUrl(uri);
      request.headers.set('Content-Type', 'application/json');
      request.headers.set('User-Agent', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36');
      
      final body = {
        'query': query,
        'context': {
          'client': {
            'clientName': 'WEB',
            'clientVersion': '2.20240501.01.00',
            'hl': 'en',
            'gl': 'US'
          }
        }
      };
      
      request.write(jsonEncode(body));
      final response = await request.close();
      if (response.statusCode != 200) {
        return [];
      }
      
      final resBody = await response.transform(utf8.decoder).join();
      final data = jsonDecode(resBody);
      
      final contents = data['contents']?['twoColumnSearchResultsRenderer']?['primaryContents']?['sectionListRenderer']?['contents'];
      if (contents == null || contents is! List) {
        return [];
      }
      
      List<dynamic>? items;
      for (var section in contents) {
        if (section['itemSectionRenderer'] != null) {
          items = section['itemSectionRenderer']['contents'];
          break;
        }
      }
      
      if (items == null) {
        return [];
      }
      
      final List<Video> videos = [];
      for (var item in items) {
        if (item['videoRenderer'] != null) {
          final videoMap = item['videoRenderer'];
          final videoIdStr = videoMap['videoId'] as String?;
          if (videoIdStr == null) continue;
          
          final title = videoMap['title']?['runs']?[0]?['text'] as String? ?? 'Unknown';
          final author = videoMap['ownerText']?['runs']?[0]?['text'] as String? ?? 'Unknown';
          final durationStr = videoMap['lengthText']?['simpleText'] as String? ?? '';
          
          Duration? duration;
          if (durationStr.isNotEmpty) {
            final parts = durationStr.split(':');
            if (parts.length == 2) {
              duration = Duration(minutes: int.tryParse(parts[0]) ?? 0, seconds: int.tryParse(parts[1]) ?? 0);
            } else if (parts.length == 3) {
              duration = Duration(hours: int.tryParse(parts[0]) ?? 0, minutes: int.tryParse(parts[1]) ?? 0, seconds: int.tryParse(parts[2]) ?? 0);
            }
          }
          
          final videoId = VideoId(videoIdStr);
          // Use a safe dummy channel ID — ChannelId() throws on free-text author names
          final channelId = ChannelId('UC${videoIdStr.padRight(22, '0')}');
          final thumbnails = ThumbnailSet(videoIdStr);
          const engagement = Engagement(0, 0, 0);
          
          final video = Video(
            videoId,
            title,
            author,
            channelId,
            DateTime.now(),
            'today',
            DateTime.now(),
            '',
            duration,
            thumbnails,
            [],
            engagement,
            false,
          );
          
          videos.add(video);
          if (videos.length >= 20) {
            break;
          }
        }
      }
      return videos;
    } catch (e) {
      debugLog('Error in InnerTube search search fallback: $e');
    } finally {
      client.close();
    }
    return [];
  }

  /// Get a Playlist and all its tracks (Using Search as a robust fallback)
  Future<Map<String, dynamic>?> getPlaylistDetails(String query) async {
    try {
      final videos = await searchTracks(query);
      if (videos.isEmpty) return null;
      
      return {
        'title': query,
        'author': 'YouTube Mix',
        'thumbnail': videos.first.thumbnails.highResUrl,
        'videos': videos,
      };
    } catch (e) {
      debugLog('Error fetching playlist via search: $e');
    }
    return null;
  }

  /// Get related tracks by searching for "<artist> <title> mix"
  Future<List<Video>> getRelatedTracks(String videoId, {String? title, String? artist}) async {
    try {
      String query;
      if (title != null && artist != null) {
        query = '$artist $title mix';
      } else {
        // Fetch the video info first to build a related query
        final video = await _yt.videos.get(videoId);
        query = '${video.author} ${video.title} mix';
      }
      final results = await searchTracks(query);
      // Skip the first result (likely the same video) and take the next batch
      return results
          .where((v) => v.id.value != videoId)
          .take(15)
          .toList();
    } catch (e) {
      debugLog('Error fetching related tracks: $e');
    }
    return [];
  }

  Future<String?> getAudioStreamUrl(String videoId) async {
    // Return from cache if available
    if (_streamCache.containsKey(videoId)) {
      debugLog('Cache hit for $videoId');
      // Move to end (LRU)
      final url = _streamCache.remove(videoId)!;
      _streamCache[videoId] = url;
      return url;
    }

    // Try high-speed direct InnerTube resolution first (takes ~150ms instead of 3-4s)
    final fastUrl = await _resolveStreamInnerTube(videoId);
    if (fastUrl != null) {
      debugLog('InnerTube fast stream resolved: $fastUrl');
      if (_streamCache.length >= _cacheMaxSize) {
        _streamCache.remove(_streamCache.keys.first);
      }
      _streamCache[videoId] = fastUrl;
      return fastUrl;
    }

    debugLog('Fast stream failed, falling back to slow YoutubeExplode client...');
    try {
      final manifest = await _yt.videos.streamsClient.getManifest(videoId);
      
      String? bestUrl;

      // Priority 1: Muxed progressive streams (extremely reliable, progressive HTTP, non-DASH)
      final muxed = manifest.muxed.sortByVideoQuality().toList();
      if (muxed.isNotEmpty) {
        bestUrl = muxed.first.url.toString();
        debugLog('Found muxed progressive stream: $bestUrl');
      }

      // Priority 2: High-bitrate audio-only m4a (AAC) stream as fallback
      if (bestUrl == null) {
        final m4aStreams = manifest.audioOnly
            .where((s) => s.container.name == 'mp4')
            .sortByBitrate()
            .toList();
        if (m4aStreams.isNotEmpty) {
          bestUrl = m4aStreams.last.url.toString();
          debugLog('Fallback to m4a audio stream: $bestUrl');
        }
      }
      
      // Priority 3: Any audio-only stream (webm/opus)
      if (bestUrl == null) {
        final anyAudio = manifest.audioOnly.sortByBitrate().toList();
        if (anyAudio.isNotEmpty) {
          bestUrl = anyAudio.last.url.toString();
          debugLog('Fallback to general audio stream: $bestUrl');
        }
      }

      if (bestUrl != null) {
        // Evict oldest if cache is full
        if (_streamCache.length >= _cacheMaxSize) {
          _streamCache.remove(_streamCache.keys.first);
        }
        _streamCache[videoId] = bestUrl;
      }

      return bestUrl;
    } catch (e) {
      debugLog('Error getting audio stream fallback: $e');
    }
    return null;
  }

  Future<String?> _resolveStreamInnerTube(String videoId) async {
    // Fire both clients in parallel - whichever resolves first wins
    final results = await Future.wait([
      _resolveStreamInnerTubeWithClient(
        videoId,
        clientName: 'ANDROID_TESTSUITE',
        clientVersion: '1.9.3',
        userAgent: 'com.google.android.youtube/19.16.34 (Linux; U; Android 11) gzip',
      ),
      _resolveStreamInnerTubeWithClient(
        videoId,
        clientName: 'ANDROID_VR',
        clientVersion: '1.37',
        userAgent: 'com.google.android.apps.youtube.vr.oculus/1.37 (Linux; U; Android 12; en_US; Quest 3; Build/SQ3A.220605.009.A1; Cronet/107.0.5284.2)',
      ),
    ]).timeout(const Duration(seconds: 4), onTimeout: () => [null, null]);

    // Return first non-null result
    return results.firstWhere((url) => url != null, orElse: () => null);
  }

  Future<String?> _resolveStreamInnerTubeWithClient(
    String videoId, {
    required String clientName,
    required String clientVersion,
    required String userAgent,
  }) async {
    final client = HttpClient()
      ..connectionTimeout = const Duration(seconds: 3);
    try {
      final uri = Uri.parse('https://www.youtube.com/youtubei/v1/player');
      final request = await client.postUrl(uri);
      request.headers.set('Content-Type', 'application/json');
      request.headers.set('User-Agent', userAgent);
      
      final body = {
        'videoId': videoId,
        'context': {
          'client': {
            'clientName': clientName,
            'clientVersion': clientVersion,
            'hl': 'en',
            'gl': 'US',
            'utcOffsetMinutes': 0
          }
        }
      };
      
      request.write(jsonEncode(body));
      final response = await request.close();
      if (response.statusCode != 200) {
        return null;
      }
      
      final resBody = await response.transform(utf8.decoder).join();
      final data = jsonDecode(resBody);
      
      final status = data['playabilityStatus']?['status'];
      if (status != 'OK') {
        return null;
      }
      
      final adaptiveFormats = data['streamingData']?['adaptiveFormats'];
      if (adaptiveFormats is List) {
        String? bestUrl;
        int maxBitrate = 0;
        
        for (var format in adaptiveFormats) {
          final mimeType = format['mimeType'] as String? ?? '';
          if (mimeType.contains('audio/')) {
            final url = format['url'] as String? ?? '';
            final bitrate = format['bitrate'] as int? ?? 0;
            if (url.isNotEmpty && bitrate > maxBitrate) {
              bestUrl = url;
              maxBitrate = bitrate;
            }
          }
        }
        return bestUrl;
      }
    } catch (e) {
      debugLog('Error in InnerTube stream resolution ($clientName): $e');
    } finally {
      client.close();
    }
    return null;
  }

  void dispose() {
    // Don't close the shared static client
  }

  void debugLog(String msg) {
    // ignore: avoid_print
    print('[YouTubeFetcher] $msg');
  }
}

File: src/youtube_fetcher.rs
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::time::{SystemTime, UNIX_EPOCH, Duration};
use reqwest::Client;
use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};
use rand::Rng;

pub struct PoTokenGenerator;

impl PoTokenGenerator {
    pub fn generate_cold_start_token(identifier: &str, client_state: &str) -> String {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        let identifier_bytes = identifier.as_bytes();
        let state_bytes = client_state.as_bytes();

        let mut rng = rand::thread_rng();
        let mut key_bytes = [0u8; 16];
        rng.fill(&mut key_bytes);

        let encrypted_id = Self::xor_encrypt(identifier_bytes, &key_bytes);
        let timestamp_bytes = Self::encode_long(timestamp);

        let mut inner_payload = Vec::new();
        inner_payload.push(0x38); // INNER_TAG
        Self::append_var_int(&mut inner_payload, state_bytes.len());
        inner_payload.extend_from_slice(state_bytes);
        inner_payload.push(0x02); // TIMESTAMP_TAG
        Self::append_var_int(&mut inner_payload, timestamp_bytes.len());
        inner_payload.extend_from_slice(&timestamp_bytes);

        let mut token_payload = Vec::new();
        token_payload.push(0x0A); // MAGIC_HEADER
        Self::append_var_int(&mut token_payload, key_bytes.len());
        token_payload.extend_from_slice(&key_bytes);
        token_payload.push(0x22); // TOKEN_VERSION
        Self::append_var_int(&mut token_payload, encrypted_id.len());
        token_payload.extend_from_slice(&encrypted_id);
        token_payload.extend_from_slice(&inner_payload);

        URL_SAFE_NO_PAD.encode(&token_payload)
    }

    fn xor_encrypt(data: &[u8], key: &[u8]) -> Vec<u8> {
        data.iter()
            .enumerate()
            .map(|(i, &b)| b ^ key[i % key.len()])
            .collect()
    }

    fn encode_long(value: u64) -> Vec<u8> {
        let mut buf = Vec::new();
        let mut v = value;
        for _ in 0..8 {
            buf.push((v & 0xFF) as u8);
            v >>= 8;
        }
        let mut len = 8;
        while len > 1 && buf[len - 1] == 0 {
            len -= 1;
        }
        buf.truncate(len);
        buf
    }

    fn append_var_int(buf: &mut Vec<u8>, value: usize) {
        let mut v = value;
        while v >= 0x80 {
            buf.push(((v & 0x7F) | 0x80) as u8);
            v >>= 7;
        }
        buf.push(v as u8);
    }
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct TrackItem {
    pub title: String,
    pub artist: String,
    pub media_id: String,
    pub thumbnail_url: String,
    pub duration_seconds: u32,
}

pub struct YouTubeFetcher {
    client: Client,
    pub innertube: crate::innertube::InnerTubeClient,
}

impl YouTubeFetcher {
    pub fn new() -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(30))
            .connect_timeout(Duration::from_secs(10))
            .build()
            .unwrap();
        let innertube = crate::innertube::InnerTubeClient::new();
        Self { client, innertube }
    }

    pub async fn get_audio_stream_url(&self, video_id: &str) -> Option<String> {
        let visitor_data = "CgtRODZlUXRVYWpQRSj0y_jSBjIKCgJTRxIEGgAgOWLfAgrcAjIwLllUPU9vdjFWVi1MSGY4Z3d5V2ltQ1BuazdBSG91NVlXZ0dYTHA0T0RCZFN2emxCNTRsblU3NjdlZURaT1Jybkg2SWZuT1I5VkRCZW55Ylh5YXIydGRmWmpWVV96NWtBeElGNnI2WHVETGt4UUpKODZDUG50T2U0N1dsdEtiUEdhanByandGOGVDYnBydXY5bkNnUGhXYktvRWxxV2VVSm01ZzFjbTdaWGVpdW5BVk16c1U2dlNCZzIyOTJiT3NkOEg4SDE2N2RGQWhCR0Q1c3IydlViSW5ldW9BQkJCcnVkWC1rYVFKcmRRbV9QVDJRODlFNm50U3NEM1JEZ3JIYWtHRDRETjk2aEZjem85OEN0NHYtbEpnRmpkcnFYMmYyV0NvUmRMek9KN2QtdTAwVVloZnlxU0NyOVczVWN3NWpTQVlVamt6YTR0cDhCT01Oblp2cFhzYzVGQQ==";
        let po_token = PoTokenGenerator::generate_cold_start_token(visitor_data, "player");

        let body = json!({
            "videoId": video_id,
            "context": {
                "client": {
                    "clientName": "ANDROID_VR",
                    "clientVersion": "1.61.48",
                    "hl": "en",
                    "gl": "US",
                    "osName": "Android",
                    "osVersion": "12",
                    "deviceMake": "Oculus",
                    "deviceModel": "Quest 3",
                    "androidSdkVersion": "32",
                    "visitorData": visitor_data,
                },
                "request": {
                    "internalExperimentFlags": [],
                    "useSsl": true
                },
                "user": {
                    "lockedSafetyMode": false
                }
            },
            "serviceIntegrityDimensions": {
                "poToken": po_token
            }
        });

        let url = "https://www.youtube.com/youtubei/v1/player?prettyPrint=false";
        let res = self.client.post(url)
            .header("Content-Type", "application/json")
            .header("User-Agent", "com.google.android.apps.youtube.vr.oculus/1.61.48 (Linux; U; Android 12; en_US; Quest 3; Build/SQ3A.220605.009.A1; Cronet/132.0.6808.3)")
            .header("Origin", "https://www.youtube.com")
            .header("Referer", "https://www.youtube.com/")
            .header("X-Origin", "https://www.youtube.com")
            .header("X-Goog-Api-Format-Version", "1")
            .header("X-YouTube-Client-Name", "28")
            .header("X-YouTube-Client-Version", "1.61.48")
            .json(&body)
            .send()
            .await;

        match res {
            Ok(response) => {
                if response.status().is_success() {
                    if let Ok(data) = response.json::<serde_json::Value>().await {
                        if let Some(status) = data["playabilityStatus"]["status"].as_str() {
                            if status != "OK" {
                                return None;
                            }
                        }

                        if let Some(adaptive_formats) = data["streamingData"]["adaptiveFormats"].as_array() {
                            let mut best_url = None;
                            let mut max_bitrate = 0;

                            for format in adaptive_formats {
                                let mime_type = format["mimeType"].as_str().unwrap_or("");
                                if mime_type.contains("audio/") {
                                    let url = format["url"].as_str().unwrap_or("");
                                    let bitrate = format["bitrate"].as_i64().unwrap_or(0);
                                    if !url.is_empty() && bitrate > max_bitrate {
                                        best_url = Some(url.to_string());
                                        max_bitrate = bitrate;
                                    }
                                }
                            }
                            return best_url;
                        }
                    }
                }
            }
            Err(_) => {}
        }
        None
    }

    pub async fn search_tracks(&self, query: &str) -> Vec<TrackItem> {
        let body = json!({
            "query": query,
            "context": {
                "client": {
                    "clientName": "WEB",
                    "clientVersion": "2.20240501.01.00",
                    "hl": "en",
                    "gl": "US"
                }
            }
        });

        let url = "https://www.youtube.com/youtubei/v1/search";
        let res = self.client.post(url)
            .header("Content-Type", "application/json")
            .header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36")
            .json(&body)
            .send()
            .await;

        let mut tracks = Vec::new();

        if let Ok(response) = res {
            if response.status().is_success() {
                if let Ok(data) = response.json::<serde_json::Value>().await {
                    if let Some(contents) = data["contents"]["twoColumnSearchResultsRenderer"]["primaryContents"]["sectionListRenderer"]["contents"].as_array() {
                        let mut items = None;
                        for section in contents {
                            if let Some(item_section) = section.get("itemSectionRenderer") {
                                if let Some(contents) = item_section.get("contents") {
                                    items = contents.as_array();
                                    break;
                                }
                            }
                        }

                        if let Some(items) = items {
                            for item in items {
                                if let Some(video) = item.get("videoRenderer") {
                                    if let Some(video_id) = video["videoId"].as_str() {
                                        let title = video["title"]["runs"][0]["text"].as_str().unwrap_or("Unknown").to_string();
                                        let artist = video["ownerText"]["runs"][0]["text"].as_str().unwrap_or("Unknown").to_string();
                                        let duration_str = video["lengthText"]["simpleText"].as_str().unwrap_or("");
                                        
                                        let mut duration_seconds = 0;
                                        if !duration_str.is_empty() {
                                            let parts: Vec<&str> = duration_str.split(':').collect();
                                            if parts.len() == 2 {
                                                let m: u32 = parts[0].parse().unwrap_or(0);
                                                let s: u32 = parts[1].parse().unwrap_or(0);
                                                duration_seconds = m * 60 + s;
                                            } else if parts.len() == 3 {
                                                let h: u32 = parts[0].parse().unwrap_or(0);
                                                let m: u32 = parts[1].parse().unwrap_or(0);
                                                let s: u32 = parts[2].parse().unwrap_or(0);
                                                duration_seconds = h * 3600 + m * 60 + s;
                                            }
                                        }

                                        let thumbnail_url = format!("https://i.ytimg.com/vi/{}/hqdefault.jpg", video_id);

                                        tracks.push(TrackItem {
                                            title,
                                            artist,
                                            media_id: video_id.to_string(),
                                            thumbnail_url,
                                            duration_seconds,
                                        });
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        tracks
    }

    #[allow(dead_code)]
    pub async fn get_related_tracks(&self, video_id: &str, title: &str, artist: &str) -> Vec<TrackItem> {
        let query = format!("{} {} mix", artist, title);
        let mut results = self.search_tracks(&query).await;
        results.retain(|t| t.media_id != video_id);
        results.truncate(15);
        results
    }
}
